### ISSUE TITLE
Path rendering issue

### ISSUE THREAD

**Issue by zpl-zak** [tdewolff/canvas#324]
Hey,

this is a follow-up on #309 , where we had an issue rendering a test SVG with some elements "missing". 

We are now up-to-date and the render does indeed look much better now, but we've noticed there's still one element rendering poorly (a blue circle with quadrant cut out).

Here is the original SVG again:
![paths](https://github.com/user-attachments/assets/383bfaea-6c7a-47a4-a9c8-3cd0a658b4c5)

And this is how the image looks like in canvas PNG renderer so far:
![path_nodes](https://github.com/user-attachments/assets/553f6e64-3cc9-4db1-b766-ca45ce0281d1)

**Comment by tdewolff** [comment]
Thanks for the issue, this should now be fixed!

**Comment by zpl-zak** [comment]
Hey, thanks for the fast response! However, it seems like there's a new issue with the rendered path, the outline is gone on some elements:

![gF6adYIVnU](https://github.com/user-attachments/assets/84a3d566-df72-4e94-9ab4-c1d1092defe8)

**Comment by tdewolff** [comment]
Yeah the fix was bad, it should be reversed. The problem is that `Path.CCW` doesn't return a valid value. It was rewritten recently and now picks the right-most segment endpoint and checks the incoming and outgoing angles of the adjacent segments. This involves a few considerations:
- The path is a single path, not multiple subpaths (this is OK, otherwise we should return a slice of bools?)
- The path does not self-intersect (this is OK, otherwise there is no reasonable result, part may be CCW and part not)
- The right-most endpoint should not be adjacent to overlapping segments, in which case we should traceback forwards and backwards until not overlapping (not implemented yet). We should check the angles and if they are equal we should traceback, but we have to be watchful whenever angles are equal but segments diverge anyways (eg. the end of a circular arc and a linear segment joining in a sharp point may have the same angle at that endpoint but diverge anyways).
- If the right-most endpoint has an adjacent segment (quadratic or cubic bezier, or elliptical arc) that goes further right still, we need to check whether that segments goes up (CCW) or down (CW). Not implemented yet and is what causes the problem above (the right-most vertex is chosen to be the bottom of the circle, and the outgoing angle for the arc is ever so slightly downwards, which causes it to be confused). If both segments go further right we pick any of the two.

**Comment by tdewolff** [comment]
Feedback:

- Point 3 above: this is rather difficult as the overlapping segments may only partly overlap for one of both segments. This probably falls in the category that paths must not self-intersect (and neither self-overlap).
- Point 4 above: this is not a problem since we make the path X-monotone first. What I've just added is the ability to distinguish between two segments with the same direction at the right-most point but with different curvatures.

The example is now fixed, can you confirm it works for you?

**Comment by zpl-zak** [comment]
Hey, the resulting image does look identical now! Thank you for the help.
